### DISCUSSION TITLE
The operator 'aten::avg_pool3d.out' is not currently implemented for the MPS device

### DISCUSSION THREAD

**Discussion by pengshunzheng** [community/community#139354] [Programming Help]
### Select Topic Area

Question

### Body

hello guys!  I used Spyder 5.4.3 to learn googlenet code，when I debug the scripts then I encountered report this error:NotImplementedError: The operator 'aten::avg_pool3d.out' is not currently implemented for the MPS device. If you want this op to be added in priority during the prototype phase of this feature, please comment on https://github.com/pytorch/pytorch/issues/77764. As a temporary fix, you can set the environment variable `PYTORCH_ENABLE_MPS_FALLBACK=1` to use the CPU as a fallback for this op. WARNING: this will be slower than running natively on MPS.  and how  I  do resolve it.  I'm a newer, hope anyone can give me a hand! thank you!

**Comment by pengshunzheng** [comment]
![Screenshot 2024-09-22 at 16 05 45](https://github.com/user-attachments/assets/9629a2c3-5a6c-422f-8d7d-885ac1f21f99)

**Comment by LiteBrite82** [comment]
Thanks for posting in the GitHub Community, @pengshunzheng ! 

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.

**Comment by tpmst** [comment]
## Use the CPU Fallback
As the error message suggests, you can enable the CPU fallback mechanism as a temporary workaround. This will cause the unsupported operations to be computed on the CPU instead of the MPS device. This is generally slower but will allow your code to run.

To enable this, you need to set the environment variable PYTORCH_ENABLE_MPS_FALLBACK=1. You can do this within your Python script before importing PyTorch, like so:
```
import os
os.environ['PYTORCH_ENABLE_MPS_FALLBACK'] = '1'

import torch
# Your code that uses PyTorch goes here
```
Alternatively, you can set this environment variable in your shell before running the Spyder or your script:
```
export PYTORCH_ENABLE_MPS_FALLBACK=1
# Now run your python script
python your_script.py
```

## Switch to CUDA
If performance is critical and you cannot wait for MPS support, consider running your computation on another device. If you have access to a Linux or Windows machine with a compatible NVIDIA GPU, using CUDA might be a viable option. On macOS, without MPS support for specific operations, you might be restricted to using the CPU only.
```
# Example to explicitly use CPU
device = torch.device('cpu')
model.to(device)
# perform your operations
```
These steps should help you bypass the NotImplementedError while using PyTorch on macOS with MPS.

**Comment by pengshunzheng** [comment]
Thank u very much! I've set the environment variable done within terminal by excuted "export PYTORCH_ENABLE_MPS_FALLBACK=1". But still report this error! Next, i'll try your approach and hope this way can solve it !thanks again!&nbsp; About the results i'll share with you!🐼🐼🐼



---Original---
From: ***@***.***&gt;
Date: Tue, Sep 24, 2024 17:01 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [community/community] The operator 'aten::avg_pool3d.out' is notcurrently implemented for the MPS device (Discussion #139354)




 
Use the CPU Fallback
 
As the error message suggests, you can enable the CPU fallback mechanism as a temporary workaround. This will cause the unsupported operations to be computed on the CPU instead of the MPS device. This is generally slower but will allow your code to run.
 
To enable this, you need to set the environment variable PYTORCH_ENABLE_MPS_FALLBACK=1. You can do this within your Python script before importing PyTorch, like so:
 import os os.environ['PYTORCH_ENABLE_MPS_FALLBACK'] = '1' import torch # Your code that uses PyTorch goes here  
Alternatively, you can set this environment variable in your shell before running the Spyder or your script:
 export PYTORCH_ENABLE_MPS_FALLBACK=1 # Now run your python script python your_script.py  
Switch to CUDA
 
If performance is critical and you cannot wait for MPS support, consider running your computation on another device. If you have access to a Linux or Windows machine with a compatible NVIDIA GPU, using CUDA might be a viable option. On macOS, without MPS support for specific operations, you might be restricted to using the CPU only.
 # Example to explicitly use CPU device = torch.device('cpu') model.to(device) # perform your operations  
These steps should help you bypass the NotImplementedError while using PyTorch on macOS with MPS.
